src/relay/backend/contrib/gcc/codegen.cc

```diff
@@ -0,0 +1,229 @@
+
+    if (GetRef<Op>(op_node) == Op::Get("add")) {
+      macro_stream << "+";
+    } else if (GetRef<Op>(op_node) == Op::Get("subtract")) {
+      macro_stream << "-";
+    } else if (GetRef<Op>(op_node) == Op::Get("multiply")) {
+      macro_stream << "*";
+    } else {
+      LOG(FATAL) << "Unrecognized op";
+    }
+
+    auto in_shape = GetShape(call->args[0]->checked_type());
+    for (size_t i = 0; i < in_shape.size(); ++i) {
+      macro_stream << ", " << in_shape[i];
+    }
+    macro_stream << ");";
+    func_decl_.push_back(macro_stream.str());
+
+    // Make function call when visiting arguments
+    bool first = true;
+    decl_stream << func_name << "(";
+    for (size_t i = 0; i < call->args.size(); ++i) {
+      VisitExpr(call->args[i]);
+      for (auto out : out_) {
+        if (!first) {
+          decl_stream << ", ";
+        }
+        first = false;
+        decl_stream << out.first;
+      }
+    }
+
+    auto type_node = call->checked_type().as<TensorTypeNode>();
+    CHECK(type_node != nullptr && runtime::TypeMatch(type_node->dtype, kDLFloat, 32))
+        << "Only support single output tensor with float type";
+    std::string out = "buf_" + std::to_string(buf_idx_++);
+    auto out_shape = GetShape(call->checked_type());
+    int out_size = 1;
+    for (size_t i = 0; i < out_shape.size(); ++i) {
+      out_size *= out_shape[i];
+    }
+    buf_stream << "float* " << out << " = (float*)std::malloc(4 * " << out_size << ");";
+    buf_decl_.push_back(buf_stream.str());
+
+    decl_stream << ", " << out << ");";
+    subgraph_body.push_back(decl_stream.str());
+
+    // Update output buffer
+    out_.clear();
+    out_.push_back({out, out_size});
+  }
+
+  /*!
+   * \brief Emit the source code that invokes gcc compatible wrappers.
+   *
+   * \return The emitted code.
+   */
+  std::string JIT() {
+    // Write function macros
+    for (auto decl : func_decl_) {
+      code_stream_ << decl << "\n";
+    }
+    return JitImpl(subgraph_id_, subgraph_args_, buf_decl_, subgraph_body, out_);
+  }
+
+ private:
+  /*! \brief The subgraph id that represents an GCC external function. */
+  std::string subgraph_id_ = "";
+  /*! \brief The index of an external function. */
+  int func_idx = 0;
+  /*! \brief The index of allocated buffers. */
+  int buf_idx_ = 0;
+  /*! \brief The arguments of a GCC compatible external function. */
+  std::vector<std::string> subgraph_args_;
+  /*! \brief The statements of a GCC compatible external function. */
+  std::vector<std::string> subgraph_body;
+  /*! \brief The declaration statements of a GCC compatible external function. */
+  std::vector<std::string> func_decl_;
+  /*! \brief The declaration statements of buffers. */
+  std::vector<std::string> buf_decl_;
+  /*! \brief The name and index pairs for output. */
+  std::vector<std::pair<std::string, int>> out_;
+};
+
+class GccCodegen : public ExternCodegenBase {
```

Comment:
GccCodgen Codes not necessarily makes sense, as the code backend is not gcc specific (works for any c compiler)

Reply:
Yes, you are right. The name doesn't make sense. How about CSourceCodegen?